### DISCUSSION TITLE
Sending device telemetry (D2C) from a cloud service without DeviceClient

### DISCUSSION THREAD

**Discussion by ghost** [Azure/azure-iot-sdk-csharp#3166] [Q&A]
Hello,
There is a service that we provide to a third-party which sends D2C messages from our system to their IoT hub.
Currently, we create a `DeviceClient` (and cache it) for each of their devices and send messages that way.

There is some scalability issues with this as more devices are added. Due to the nature of the device client, having multiple instances is not supported so scaling this service horizontally is not possible.

Is there a more "lightweight" or scalable way to send D2C messages without spinning up device client instances? 
I'm thinking potentially with something like `ServiceClient` or another class?

**Comment by drwill-ms** [comment]
You might want to use AMQP connection pooling (multiplexing).

Initialize your `DeviceClient` instances with connection settings like these:

```csharp
            var transportSettings = new ITransportSettings[]
            {
                new AmqpTransportSettings(TransportType.Amqp_Tcp_Only)
                {
                    AmqpConnectionPoolSettings = new AmqpConnectionPoolSettings
                    {
                        MaxPoolSize = 100, // pick a value that works for you, default is 100
                        Pooling = true,
                    }
                }
            };
```

I'm sorry to say we don't have a sample for this. We should add one but we won't be able to get to it in the near future.

**Reply by ghost** [reply]
Sorry I didn't mention it in the original post, but we're not having issues with port exhaustion / pooling. Our current setup has these settings defined already.

The main concern is being able to scale this service horizontally (multiple instances) and I was wondering if there is alternative methods to sending D2C messages without `DeviceClient`. Possibly something that doesn't require a single instance per device and uses less resources.

I understand that this may be out-of-scope for an IoT library but was wondering if the SDK team had any insights / recommendations.

**Reply by drwill-ms** [reply]
Using our SDK, the only option to send telemetry (D2C) messages is using a `DeviceClient` instance. The pooling is to help scale out on the resource that hosts are most likely going to run out of (connections). What is the concern about having a `DeviceClient` instance per device that the app is simulating? Memory usage?

**Reply by ghost** [reply]
Primarily memory usage, yeah. 
We've considered creating/disposing a `DeviceClient` on each request but I'm not sure it would perform well or be supported as the amount of requests can range from 120/sec to ~200/sec. We do this in another web application but the request volume is much lower.

It's web app so we'd be forced to scale up (add more resources) right now which is more costly/restrictive than scaling out (adding more servers)

**Reply by drwill-ms** [reply]
We don't think there is a lot of memory overhead for device clients but perhaps we aren't looking at the scale you are working at. I think for using temporary device client instances it would depend on how much later you'd want to represent that device in another operation.
